src/audio_core/time_stretch.cpp

```diff
@@ -62,8 +65,41 @@ std::size_t TimeStretcher::Process(const s16* in, std::size_t num_in, s16* out,
     LOG_TRACE(Audio, "{:5}/{:5} ratio:{:0.6f} backlog:{:0.6f}", num_in, num_out, stretch_ratio,
               backlog_fullness);
 
-    sound_touch->putSamples(in, static_cast<u32>(num_in));
-    return sound_touch->receiveSamples(out, static_cast<u32>(num_out));
+    if constexpr (std::is_floating_point<soundtouch::SAMPLETYPE>()) {
+        // SoundTouch library on most systems expects float samples
+        // use this vector to store input if soundtouch::SAMPLETYPE is a float
+        std::vector<soundtouch::SAMPLETYPE> float_in(2 * num_in);
+        std::vector<soundtouch::SAMPLETYPE> float_out(2 * num_out);
+
+        for (std::size_t i = 0; i < (2 * num_in); i++) {
+            // Conventional integer PCM uses a range of -32768 to 32767,
+            // but float samples use -1 to 1
+            // As a result we need to scale sample values during conversion
+            float temp = static_cast<float>(in[i]) / std::numeric_limits<s16>::max();
+            float_in[i] = static_cast<soundtouch::SAMPLETYPE>(temp);
+        }
+
+        sound_touch->putSamples(float_in.data(), static_cast<u32>(num_in));
+
+        std::size_t samples_received =
+            sound_touch->receiveSamples(float_out.data(), static_cast<u32>(num_out));
+
+        // Converting output samples back to shorts so we can use them
+        for (std::size_t i = 0; i < (2 * num_out); i++) {
+            s16 temp = static_cast<s16>(float_out[i] * std::numeric_limits<s16>::max());
+            out[i] = temp;
+        }
+
+        return samples_received;
+    } else if (std::is_same<soundtouch::SAMPLETYPE, s16>()) {
+        sound_touch->putSamples(in, static_cast<u32>(num_in));
+        return sound_touch->receiveSamples(out, static_cast<u32>(num_out));
```

Comment:
```suggestion
        sound_touch->putSamples(reinterpret_cast<soundtouch::SAMPLETYPE*>(in), static_cast<u32>(num_in));
        return sound_touch->receiveSamples(reinterpret_cast<soundtouch::SAMPLETYPE*>(out), static_cast<u32>(num_out));
```
Should work I guess... would have been nice if soundtouch provided a better interface

Reply:
> would have been nice if soundtouch provided a better interface

Glad it's not just me ;)

Anyway the cast on line 95 needs to be `const` or else it complains about casting away qualifiers. Apart from that it works perfectly. Good work on the `reinterpret_cast`. I never would have thought of that.

Reply:
Fixed